derives/src/derive_labelled_generic.rs

```diff
@@ -185,9 +215,27 @@ fn build_field_constr_for(field: &Field) -> impl ToTokens {
 /// Assumes that there are bindings in the immediate environment with those names that
 /// are bound to Field values.
 ///
-/// The opposite of build_labelled_hcons_constr
-fn build_new_labelled_struct_constr(struct_name: &Ident, bindnames: &Vec<Ident>) -> impl ToTokens {
+/// The opposite of build_labelled_hcons_constr for named structs
+fn build_new_labelled_struct_constr(
+    struct_name: &Ident,
+    bindnames: &Vec<Ident>,
+) -> Box<dyn ToTokens> {
     let cloned_bind1 = bindnames.clone();
     let cloned_bind2 = bindnames.clone();
-    quote! { #struct_name { #(#cloned_bind1: #cloned_bind2.value),* } }
+    Box::new(quote! { #struct_name { #(#cloned_bind1: #cloned_bind2.value),* } })
+}
+
+/// Given a tuple struct name, and a number of Idents that act as accessors
+/// names, returns an AST representing how to construct said tuple struct.
+///
+/// Assumes that there are bindings in the immediate environment with those names that
+/// are bound to Field values.
+///
+/// The opposite of build_labelled_hcons_constr for yuple structs
+fn build_new_labelled_tuple_struct_constr(
+    struct_name: &Ident,
+    bindnames: &Vec<Ident>,
+) -> Box<dyn ToTokens> {
```

Comment:
Is there a reason why `impl ToTokens` doesn't work here?

Reply:
These return types have to unify in the `match` statement when they're called, and `impl Trait` return values cannot unify (except with themselves).

By returning a trait object, these two return types can be used interchangeably.

Reply:
Ah, the match is somewhere else... OK, let's move the `Box::new(...)` to that place instead and write `-> impl Trait` here so that the reason for the dynamic dispatch is clearer and so that it is done where it is actually needed.